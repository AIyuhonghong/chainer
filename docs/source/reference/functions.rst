Standard Function implementations
=================================

.. currentmodule:: chainer.functions

Parameterized function classes
------------------------------
.. autoclass:: BatchNormalization
.. autoclass:: Convolution2D
.. autoclass:: EmbedID
.. autoclass:: Linear
.. autoclass:: Parameter
.. autoclass:: PReLU

Array manipulation functions
----------------------------
.. autofunction:: concat
.. autofunction:: copy
.. autofunction:: dropout
.. autofunction:: identity

Activation functions
--------------------
.. autofunction:: leaky_relu
.. autofunction:: lstm
.. autofunction:: relu
.. autofunction:: sigmoid
.. autofunction:: softmax
.. autofunction:: tanh

Pooling functions
-----------------
.. autofunction:: average_pooling_2d
.. autofunction:: max_pooling_2d

Loss, evaluation and aggregation
--------------------------------
.. autofunction:: accuracy
.. autofunction:: mean_squared_error
.. autofunction:: softmax_cross_entropy
.. autofunction:: sum
