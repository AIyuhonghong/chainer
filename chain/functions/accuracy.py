import numpy
import pycuda.gpuarray as gpuarray
from chain import Function

class Accuracy(Function):
    """Compute accuracy within minibatch."""

    def forward_cpu(self, inputs):
        y, t = inputs
        pred = y.argmax(axis=1)
        return (pred == t).mean(keepdims=True),

    def forward_gpu(self, inputs):
        # Fallback to CPU
        # TODO(beam2d): Pure GPU version
        accuracy = self.forward_cpu((a.get() for a in inputs))
        return gpuarray.to_gpu(accuracy),


def accuracy(y, t):
    return Accuracy()(y, t)
