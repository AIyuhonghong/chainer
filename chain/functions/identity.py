from chain import Function

class Identity(Function):
    """Identity function."""

    def forward(self, xs):
        return xs

    def backward(self, xs, gys):
        return gys

def identity(*inputs):
    return Identity()(*inputs)
