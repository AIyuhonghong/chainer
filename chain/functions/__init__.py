# Non-parameterized functions
from accuracy import accuracy
from dropout  import dropout
from identity import identity
from relu     import relu
from sigmoid  import sigmoid
from softmax  import softmax
from softmax_cross_entropy import softmax_cross_entropy
from tanh     import tanh

# Parameterized layers
from embed_id import EmbedID
from linear   import Linear
