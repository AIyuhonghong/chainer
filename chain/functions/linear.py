import math
import numpy
import pycuda.gpuarray as gpuarray
import scikits.cuda.linalg as culinalg
import scikits.cuda.misc as cumisc

from chain import Function

class Linear(Function):
    """Implementation of fully-connected layer."""

    parameter_names = ('W', 'b')
    gradient_names  = ('gW', 'gb')

    def __init__(self, in_size, out_size, wscale=1):
        self.W = numpy.random.normal(
            0, wscale * math.sqrt(1. / in_size),
            (out_size, in_size)).astype(numpy.float32)
        self.b = numpy.zeros((out_size,), dtype=numpy.float32)

        self.gW = numpy.empty_like(self.W)
        self.gb = numpy.empty_like(self.b)

    def forward_cpu(self, x):
        return x[0].dot(self.W.T) + self.b,

    def forward_gpu(self, x):
        Wx = culinalg.dot(x[0], self.W, transb='T')
        return cumisc.add_matvec(Wx, self.b, axis=1),

    def backward_cpu(self, x, gy):
        self.gW += gy[0].T.dot(x[0])
        self.gb += gy[0].sum(0)
        return gy[0].dot(self.W),

    def backward_gpu(self, x, gy):
        culinalg.add_dot(gy[0], x[0], self.gW, transa='T')
        self.gb += cumisc.sum(gy[0], 0)
        return culinalg.dot(gy[0], self.W),
