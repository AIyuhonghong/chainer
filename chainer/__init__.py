import pkg_resources

from chainer import function
from chainer import function_set
from chainer.functions import basic_math
from chainer import optimizer
from chainer import serializer
from chainer import variable


__version__ = pkg_resources.get_distribution('chainer').version

AbstractSerializer = chainer.AbstractSerializer
Deserializer = chainer.Deserializer
Function = function.Function
FunctionSet = function_set.FunctionSet
Optimizer = optimizer.Optimizer
Serializer = chainer.Serializer
Variable = variable.Variable

basic_math.install_variable_arithmetics()
