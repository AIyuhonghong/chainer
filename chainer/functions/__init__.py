# Non-parameterized functions
from accuracy   import accuracy
from concat     import concat
from dropout    import dropout
from identity   import identity
from lstm       import lstm
from pooling_2d import average_pooling_2d, max_pooling_2d
from relu       import relu
from sigmoid    import sigmoid
from softmax    import softmax
from softmax_cross_entropy import softmax_cross_entropy
from sum        import sum
from tanh       import tanh

# Parameterized layers
from convolution_2d import Convolution2D
from embed_id       import EmbedID
from inception      import Inception
from linear         import Linear
