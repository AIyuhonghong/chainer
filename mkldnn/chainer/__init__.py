import os
import numpy

from mkldnn.mdarray import mdarray

from chainer import variable
from chainer import configuration

from chainer.configuration import config
from chainer.configuration import global_config

# export CHAINER_ENABLE_COSIM=0
global_config.cosim = bool(int(os.environ.get('CHAINER_ENABLE_COSIM', '0')))
# export CHAINER_ENABLE_COSIM_CONTINUE=0
global_config.cosim_continue = bool(int(os.environ.get('CHAINER_ENABLE_COSIM_CONTINUE', '0')))

if global_config.cosim is True:
    variable.RANK_START = 1


def is_cosim_continue():
    """Get the cosim continue mode

    Returns:
        bool: Return ``True`` if continue after a failure.
    """
    return config.cosim_continue


def is_cosim():
    """Get the cosim mode.

    Returns:
        bool: Return ``True`` if Chainer is in cosim mode.
    """
    return config.cosim


def plain_array(params):
    assert isinstance(params, tuple) \
            or isinstance(params, list) \
            or isinstance(params, mdarray) \
            or isinstance(params, numpy.ndarray) \
            or isinstance(params, variable.Variable)

    _params = ()

    if isinstance(params, variable.Variable):
        return (numpy.array(params.data), )
    elif isinstance(params, numpy.ndarray):
        return (params, )
    elif isinstance(params, mdarray):
        return (numpy.array(params), )

    for p in params:
        if isinstance(p, variable.Variable):
            p = numpy.array(p.data)
        if isinstance(p, mdarray):
            _params += (numpy.array(p), )
        else:
            _params += (p, )

    return _params

