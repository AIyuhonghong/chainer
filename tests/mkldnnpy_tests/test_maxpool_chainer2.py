import chainer
import chainer.functions as F
import chainer.links as L
import numpy as np
import math

#input = [   [ 1, 2, 3 ],
#            [ 4, 5, 6 ],
#            [ 7, 8, 9 ] ]

x = np.zeros((1, 1, 4, 4), dtype=np.float32)

for i in range(1):
    for j in range(1):
        for k in range(4):
            for l in range(4):
                x[i, j, k, l] = math.sin(i+j*2+k*3+l*4)
                                        # break symmetry
gy = np.ndarray((1, 1, 4, 4), dtype=np.float32)
gy.fill(0.01)

x=x,
gy=gy,

print "first pass"
print "x="
print x[0]
f = F.MaxPooling2D(3, stride=1, pad =1, use_cudnn=False)
y=f.forward_cpu(x)
print "y="
print y[0]
gx=f.backward_cpu(x, gy)
print "gx="
print gx[0]


print "second pass"
print "x="
print x[0]
y=f.forward_cpu(x)
print "y="
print y[0]
gx = f.backward_cpu(x, gy)
print "gx="
print gx[0]
