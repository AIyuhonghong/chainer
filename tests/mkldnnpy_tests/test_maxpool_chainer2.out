first pass
x=
[[[[ 0.         -0.7568025   0.98935825 -0.53657293]
   [ 0.14112     0.65698659 -0.99999022  0.65028787]
   [-0.27941549 -0.54402113  0.99060738 -0.75098723]
   [ 0.41211849  0.42016703 -0.96139747  0.83665562]]]]
y=
[[[[ 0.65698659  0.98935825  0.98935825  0.98935825]
   [ 0.65698659  0.99060738  0.99060738  0.99060738]
   [ 0.65698659  0.99060738  0.99060738  0.99060738]
   [ 0.42016703  0.99060738  0.99060738  0.99060738]]]]
gx=
[[[[ 0.          0.          0.03        0.        ]
   [ 0.          0.03        0.          0.        ]
   [ 0.          0.          0.08999999  0.        ]
   [ 0.          0.01        0.          0.        ]]]]
second pass
x=
[[[[ 0.         -0.7568025   0.98935825 -0.53657293]
   [ 0.14112     0.65698659 -0.99999022  0.65028787]
   [-0.27941549 -0.54402113  0.99060738 -0.75098723]
   [ 0.41211849  0.42016703 -0.96139747  0.83665562]]]]
y=
[[[[ 0.65698659  0.98935825  0.98935825  0.98935825]
   [ 0.65698659  0.99060738  0.99060738  0.99060738]
   [ 0.65698659  0.99060738  0.99060738  0.99060738]
   [ 0.42016703  0.99060738  0.99060738  0.99060738]]]]
gx=
[[[[ 0.          0.          0.03        0.        ]
   [ 0.          0.03        0.          0.        ]
   [ 0.          0.          0.08999999  0.        ]
   [ 0.          0.01        0.          0.        ]]]]
